File: Languages/cpp/algorithms-dsa/leetcode/union-find/redundant_connection.cpp
/**
 * LeetCode 684: Redundant Connection
 * Topic: Union-Find
 * Difficulty: Medium
 *
 * Given a graph that was a tree plus one extra edge, find the redundant edge.
 * Use Union-Find to detect the first edge that creates a cycle.
 * Time: O(n * α(n)) ≈ O(n), Space: O(n)
 */
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

class UnionFind {
    vector<int> parent, rank_;
public:
    UnionFind(int n) : parent(n + 1), rank_(n + 1, 0) {
        for (int i = 0; i <= n; i++) parent[i] = i;
    }
    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    bool unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false; // cycle detected
        if (rank_[px] < rank_[py]) swap(px, py);
        parent[py] = px;
        if (rank_[px] == rank_[py]) rank_[px]++;
        return true;
    }
};

vector<int> findRedundantConnection(vector<vector<int>>& edges) {
    int n = edges.size();
    UnionFind uf(n);
    for (auto& e : edges) {
        if (!uf.unite(e[0], e[1])) return e;
    }
    return {};
}

int main() {
    vector<vector<int>> e1 = {{1,2},{1,3},{2,3}};
    assert(findRedundantConnection(e1) == vector<int>({2,3}));

    vector<vector<int>> e2 = {{1,2},{2,3},{3,4},{1,4},{1,5}};
    assert(findRedundantConnection(e2) == vector<int>({1,4}));

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/union-find/redundant_connection.py
"""
LeetCode #684 - Redundant Connection
Topic: Union-Find
Difficulty: Medium

Given a graph that started as a tree with n nodes and had one extra edge added,
find the edge that can be removed to make it a tree again. If multiple answers,
return the one that appears last in the input.

Approach: Union-Find. Process edges one by one; the first edge that connects
two already-connected nodes forms the cycle and is our answer.

Time Complexity: O(n * α(n)) ≈ O(n)
Space Complexity: O(n)
"""


class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False  # cycle detected
        if self.rank[px] < self.rank[py]:
            px, py = py, px
        self.parent[py] = px
        if self.rank[px] == self.rank[py]:
            self.rank[px] += 1
        return True


def find_redundant_connection(edges: list[list[int]]) -> list[int]:
    n = len(edges)
    uf = UnionFind(n + 1)  # 1-indexed nodes
    for u, v in edges:
        if not uf.union(u, v):
            return [u, v]
    return []


if __name__ == "__main__":
    assert find_redundant_connection([[1, 2], [1, 3], [2, 3]]) == [2, 3]
    assert find_redundant_connection([[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]]) == [1, 4]
    assert find_redundant_connection([[1, 2], [1, 3], [1, 4], [3, 4]]) == [3, 4]
    print("All tests passed!")
